### PR TITLE
Views (community)

### PR DESCRIPTION
**Pull request by Polymorphe57**

### CODE REVIEW

File: addons/wowl/static/src/types.ts [hunk 1]
```diff
@@ -243,20 +244,23 @@ export interface SystrayItem {
...
+  | "datetime"
+  | "selection";
 
 export interface FieldDefinition {
   relation?: string;
   relation_field?: string;
   string: string;
   type: FieldType;
   default?: any;
+  selection?: any[][];
```
Comment (Polymorphe57) [inline addons/wowl/static/src/types.ts:263]: not sure

File: addons/wowl/static/src/views/favorite_menu/favorite_menu.xml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <t t-name="wowl.FavoriteMenu" owl="1">
+        <Dropdown t-on-dropdown-item-selected.stop="onDropdownItemSelected" class="o_favorite_menu" togglerClass="'btn btn-secondary'">
+            <Dialog t-if="state.toDelete" title="'Warning'" t-on-dialog-closed="emptyState">
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/favorite_menu/favorite_menu.xml]: size is not good

File: addons/wowl/static/src/views/view.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import { ControlPanelSubTemplates } from "./abstract_controller";
+import { Model, ModelParams } from "./model";
+import { FavoriteMenu } from "./favorite_menu/favorite_menu";
+import { Dialog } from "../components/dialog/dialog";
+import { Dropdown } from "../components/dropdown/dropdown";
+import { DropdownItem } from "../components/dropdown/dropdown_item";
+const { useSubEnv } = hooks;
+
+export interface Fields {
+  [fieldName: string]: FieldDefinition;
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view.ts]: FieldDefinition could be a union of "field type" in order to have better control and do not have optional parameters (similar to too loose rng validation)

File: addons/wowl/static/tests/helpers/control_panel.ts [hunk 4]
```diff
@@ -1,6 +1,10 @@
 import { Component } from "@odoo/owl";
+import { ViewType } from "../../src/types";
+import { getLegacy } from "./legacy";
 import { click } from "./utility";
 
+getLegacy;
```
Comment (Polymorphe57) [inline addons/wowl/static/tests/helpers/control_panel.ts]: Warning warning!

File: addons/wowl/static/tests/action_manager/window_action_tests.ts [hunk 5]
```diff
@@ -1581,15 +1592,11 @@ QUnit.module("ActionManager", (hooks) => {
...
-    testUtils.mock.patch(ListController, {
-      getOwnedQueryParams: function () {
-        return {
-          context: {
-            shouldBeInFilterContext: true,
-          },
-        };
-      },
-    });
+    const onSaveParams = ListView.onSaveParams;
```
Comment (Polymorphe57) [inline addons/wowl/static/tests/action_manager/window_action_tests.ts]: Warning warning!

File: addons/wowl/static/src/views/graph/graph_view.xml [hunk 6]
```diff
@@ -0,0 +1,128 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <div t-name="wowl.NoContentHelper" class="o_view_nocontent" role="alert" owl="1">
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/graph/graph_view.xml]: should be moved somewhere else

File: addons/wowl/static/src/views/graph/graph_view.xml [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            </p>
+            <p>
+                <t t-if="description" t-esc="description"/>
+                <t t-else="1">Try to add some records, or make sure that there is no
+                    active filter in the search bar.</t>
+            </p>
+        </div>
+    </div>
+
+    <div t-name="wowl.ActionHelper" class="o_view_nocontent" owl="1">
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/graph/graph_view.xml]: should be moved somewhere else

File: addons/wowl/static/src/views/graph/types.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  string: string;
+  type: FieldType;
+  default?: any;
+  selection?: [any, string][];
+  store?: boolean;
+  sortable?: boolean;
+  context?: Context; // does it exists really?
+}
+
+export interface Fields {
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/graph/types.ts]: should be moved somewhere else

File: addons/wowl/static/src/views/view_utils/favorite_menu/custom_favorite_item.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+  _notificationService: NotificationService = useService("notifications");
+
+  descriptionRef = useRef("description");
+  useByDefaultId: string;
+  shareAllUsersId: string;
+
+  interactive: boolean = true;
+  state: State = {
+    description: /** this.env.action.name ||*/ "",
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view_utils/favorite_menu/custom_favorite_item.ts]: don't forget this

File: addons/wowl/static/src/views/view_utils/favorite_menu/custom_favorite_item.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  };
+
+  constructor() {
+    super(...arguments);
+
+    const favId = favoriteId++;
+    this.useByDefaultId = `o_favorite_use_by_default_${favId}`;
+    this.shareAllUsersId = `o_favorite_share_all_users_${favId}`;
+
+    // useAutofocus();
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view_utils/favorite_menu/custom_favorite_item.ts]: don't forget this

File: addons/wowl/static/src/views/view_utils/favorite_menu/favorite_menu.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    userId: number | false;
+  };
+}
+
+export class FavoriteMenu extends Component<{ searchModel?: SearchModel }, ViewEnv> {
+  static registry = new Registry();
+  static components = { Dialog, Dropdown, DropdownItem };
+  static template = "wowl.FavoriteMenu";
+
+  searchModel: SearchModel = this.props.searchModel || this.env.searchModel;
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view_utils/favorite_menu/favorite_menu.ts]: remove part "this.env.searchModel" (same things in similar components)

File: addons/wowl/static/tests/action_manager/window_action_tests.ts (hunk 5 above)
Comment (Polymorphe57) [reply addons/wowl/static/tests/action_manager/window_action_tests.ts]: move the test somewhere else

File: addons/wowl/static/src/views/view_utils/view_loader.ts [hunk 12]
```diff
@@ -31,6 +34,17 @@ export class ViewLoader extends Component<any, any> {
     // extract props from action
     this.viewProps.limit = this.props.action.limit;
 
+    if (this.props.views.find((v: [number, string]) => v[1] === "search")) {
+      this.viewProps.searchModelParams = await processSearchViewDescription(
+        viewDescriptions.search,
+        this._modelService,
+        this.props.context
+      );
+      Object.assign(this.viewProps.searchModelParams, {
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view_utils/view_loader.ts]: remove

File: addons/wowl/static/src/views/graph/graph_view.ts [hunk 13]
```diff
@@ -355,11 +389,21 @@ export class GraphView extends Component<GraphViewProps, OdooEnv> {
   }
 
   onMeasureSelected(ev: OwlEvent<{ payload: { measure: string } }>) {
-    this.model.activeMeasure = ev.detail.payload.measure;
+    // Not good!!!!!!! Don't know where activeMeasure, mode, order should go!!! in model?
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/graph/graph_view.ts]: Warning warning!

File: addons/wowl/static/src/views/graph/graph_model.ts [hunk 14]
```diff
@@ -14,6 +13,28 @@ const NO_CONTENT_HELPER_DESCRIPTION = _lt(
   "Pie chart cannot mix positive and negative numbers. Try to change your domain to only display positive results"
 );
 
+export interface GraphModelParams {
+  modelName: string;
+  fields: Fields;
+}
+
+export interface GraphModelLoadParams {
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/graph/graph_model.ts]: We end up with the same thing as before ModelParams and LoadParams,... But we could move the ModelParams to LoadParams

File: addons/wowl/static/src/views/graph/graph_view.xml [hunk 15]
```diff
@@ -109,7 +109,7 @@
                 </div>
             </t>
             <t t-slot="default"> <!-- t-set-slot does not work here ??? -->
-                <label t-if="title" t-esc="title"/>
+                <label t-if="props.title" t-esc="props.title"/>
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/graph/graph_view.xml]: props.title --> title

File: addons/wowl/static/tests/views/graph/graph_view_tests.ts [hunk 16]
```diff
@@ -167,38 +189,203 @@ QUnit.module(
...
+      assert.expect(8);
+      const props: GraphProps = { model: "foo" };
+      const graph = await mount(GraphView, { env, target, props });
+      assert.containsOnce(graph.el!, "div.o_graph_canvas_container canvas");
+      assert.strictEqual(
+        graph.state.activeMeasure,
+        DEFAULT_MEASURE,
+        `the active measure should be "${DEFAULT_MEASURE}" by default`
+      );
+      assert.strictEqual(graph.state.mode, "bar", "should be in bar chart mode by default");
```
Comment (Polymorphe57) [inline addons/wowl/static/tests/views/graph/graph_view_tests.ts]: don't like this kind of assertions

File: addons/wowl/doc/services/user.md [hunk 17]
```diff
@@ -10,11 +10,10 @@ The `user` service is a very simple service, that aims to just keep track of a
 few important values related to the current user. It simply provides an object
 with a few keys:
 
-- `allowed_companies ([number, string][])`: the list of companies that can be
-  accessed by the user. Each element is a pair `id, name`
+- `allowed_companies ({[id: number] : {id: number, name: string} })`: the object of companies that can be accessed by the user. Each value is a pair `id, name`
```
Comment (Polymorphe57) [inline addons/wowl/doc/services/user.md]: review explanations

File: addons/wowl/static/src/services/user_service.js [hunk 18]
```diff
@@ -2,22 +2,16 @@
...
+
+  let allowedCompanies = cidsFromHash || [];
+  const allowedCompaniesFromSession = user_companies.allowed_companies;
   const notReallyAllowedCompanies = allowedCompanies.filter(
-    (id) => !allowedCompaniesFromSession.includes(id)
+    (id) => !(id in allowedCompaniesFromSession)
   );
   if (!allowedCompanies.length || notReallyAllowedCompanies.length) {
-    allowedCompanies = [user_companies.current_company[0]];
+    allowedCompanies = [user_companies.current_company];
```
Comment (Polymorphe57) [inline addons/wowl/static/src/services/user_service.js]: hum something is weird

File: addons/wowl/static/src/services/user_service.js [hunk 19]
```diff
@@ -58,7 +60,7 @@ export const userService = {
       isAdmin: is_admin,
       partnerId: partner_id,
       allowed_companies: user_companies.allowed_companies,
-      current_company: user_companies.current_company,
+      current_company: user_companies.allowed_companies[allowedCompanies[0]],
```
Comment (Polymorphe57) [inline addons/wowl/static/src/services/user_service.js]: here we select the first object in user_companies.allowed_companies...

Comment (Polymorphe57) [reply addons/wowl/static/src/services/user_service.js]: do we want the id instead? What is expected really?

File: addons/wowl/static/src/services/model_service.js [hunk 20]
```diff
@@ -40,7 +41,14 @@ function readGroup(rpc, env, model) {
     if (options.limit) {
       kwargs.limit = options.limit;
     }
-    return callModel(rpc, env, model)("web_read_group", [], kwargs);
+    const result = await callModel(rpc, env, model)("web_read_group", [], kwargs);
+    if (result && result.length === 1) {
+      // hum, no result really?
+      // API correction of read_group
+      result.groups[0].__domain = result.groups[0].__domain || [];
```
Comment (Polymorphe57) [inline addons/wowl/static/src/services/model_service.js]: to do? or make another read_group function that makes some processing for the reporting view?

File: addons/wowl/static/src/views/view_service.js [hunk 21]
```diff
@@ -1,7 +1,74 @@
 /** @odoo-module **/
-
 import { serviceRegistry } from "../webclient/service_registry";
 
+// How do we manage interfaces, types,... ?
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view_service.js]: clean/keep/abandon types/interfaces --> what policy?

File: addons/wowl/static/tests/helpers/index.js [hunk 22]
```diff
@@ -60,6 +60,8 @@ export { makeFakeUserService, makeFakeRPCService, makeMockXHR, makeMockFetch } f
 export {
   click,
   getFixture,
+  getTestServiceRegistry,
```
Comment (Polymorphe57) [inline addons/wowl/static/tests/helpers/index.js]: makeTestServiceRegistry,..

File: addons/wowl/static/tests/helpers/mocks.js [hunk 23]
```diff
@@ -82,7 +84,10 @@ export function makeFakeUserService(values) {
 }*/
 
 function buildMockRPC(mockRPC) {
-  return async (...args) => {
+  return async function (...args) {
+    if (this instanceof Component && this.__owl__.isDestroyed) {
```
Comment (Polymorphe57) [inline addons/wowl/static/tests/helpers/mocks.js]: && this.__owl__.status === 5

File: addons/wowl/static/src/services/model_service.js (hunk 20 above)
Comment (Polymorphe57) [reply addons/wowl/static/src/services/model_service.js]: not to do!

Comment (Polymorphe57) [reply addons/wowl/static/src/services/model_service.js]: create other function --> find name + implementation

File: addons/wowl/static/src/services/model_service.js [hunk 24]
```diff
@@ -21,14 +21,15 @@ function write(rpc, env, model) {
...
-  return (domain, fields, groupby, options = {}, ctx = {}) => {
+  return async (domain, fields, groupby, options = {}, ctx = {}) => {
     const kwargs = {
       domain,
       groupby,
       fields,
       context: ctx,
     };
-    if (options.lazy) {
+    if ("lazy" in options) {
```
Comment (Polymorphe57) [inline addons/wowl/static/src/services/model_service.js]: ok

File: addons/wowl/static/src/views/view_service.js (hunk 21 above)
Comment (Polymorphe57) [reply addons/wowl/static/src/views/view_service.js]: clean types

File: addons/wowl/static/src/core/hooks.js [hunk 25]
```diff
@@ -15,9 +15,14 @@ export function useService(serviceName) {
   const component = Component.current;
   const env = component.env;
   const service = env.services[serviceName];
-  if (!service) {
+  if (service === undefined) {
```
Comment (Polymorphe57) [inline addons/wowl/static/src/core/hooks.js]: if (serviceNane in env.services) ...
+ tests

Comment (Polymorphe57) [reply addons/wowl/static/src/core/hooks.js]: static/test/hook/...

File: addons/wowl/static/src/actions/action_service.js [hunk 26]
```diff
@@ -161,11 +162,13 @@ function makeActionManager(env) {
           multiRecord: v.multiRecord,
         };
       });
+    // FIXME: props should be cleaned
     const props = Object.assign({}, _getActionProps(action), {
       context: action.context,
       domain: action.domain || [],
       model: action.res_model,
       type: view.type,
+      View: view,
```
Comment (Polymorphe57) [inline addons/wowl/static/src/actions/action_service.js]: we should check that part

Comment (Polymorphe57) [reply addons/wowl/static/src/actions/action_service.js]: and all changes here: something seems weird

File: addons/wowl/static/src/views/view_utils/misc.js [hunk 27]
```diff
@@ -0,0 +1,21 @@
+/** @odoo-module **/
+
+export const VIEW_DEFAULT_PROPS = {
+  fields: {},
+  isEmbedded: false,
+  isSample: false,
+};
+
+export const VIEW_PROPS = {
+  __beforeLeave__: { type: Function, optional: 1 },
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view_utils/misc.js]: not nice to have this here

File: addons/wowl/static/src/views/view_utils/misc.js [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+  __exportState__: { type: Function, optional: 1 }, // hum hum for export ...
+  actionName: { type: String, optional: 1 },
+  fields: { type: Object, elements: Object }, // more precision on elements...
+  model: String,
+  isEmbedded: Boolean,
```
Comment (Polymorphe57) [inline addons/wowl/static/src/views/view_utils/misc.js]: isEmbedded is no more a good concept

File: addons/web/static/src/actions/action_service.js [hunk 29]
```diff
@@ -226,7 +226,13 @@ function makeActionManager(env) {
     }
     if (action.controllers[view.type]) {
       // this controller has already been used, re-import its exported state
-      props.state = action.controllers[view.type].exportedState;
+      if ("exportedState" in action.controllers[view.type]) {
+        debugger
```
Comment (Polymorphe57) [inline addons/web/static/src/actions/action_service.js]: bug here --> willUnmount to late --> use event "ACTION_MANAGER:EXPORT_CONTROLLER_STATE"

File: addons/web/static/src/core/l10n/numbers.js [hunk 30]
```diff
@@ -91,6 +91,101 @@ export function parseFloat(value) {
...
+/**
+ * Parse a string containing a float and unconvert it with a conversion factor
+ *
+ * @param {string} value the string to be parsed
+ * @param {Object} [options]
+ * @param {number} [options.factor=1.0] conversion factor
+ * @returns float
+ * @throws Error if the value can't be parsed properly
+ */
+export function parseFloatFactor(value, options) {
```
Comment (ged-odoo) [inline addons/web/static/src/core/l10n/numbers.js]: remove this

Comment (Polymorphe57) [inline addons/web/static/src/core/l10n/numbers.js]: remove

File: addons/project/views/rating_views.xml [hunk 31]
```diff
@@ -66,8 +66,12 @@
...
         <field name="mode">primary</field>
         <field name="arch" type="xml">
-            <xpath expr="//graph" position="attributes">
-                <attribute name="js_class">project_rating_graph</attribute>
+            <xpath expr="//field[@name='rating']" position="attributes">
+                <attribute name="string">Rating Value (/5)</attribute>
+            </xpath>
+            <xpath expr="//graph" position="inside">
+                <field name="res_id" type="measure" string="Task"/>
+                <field name="parent_res_id" type="measure" string="Project"/>
```
Comment (Polymorphe57) [inline addons/project/views/rating_views.xml]: check that initial measure is preserved

**Pull request closed without merging**